### PR TITLE
feat: initial code push

### PR DESCRIPTION
**Pull request by eneemann**
Here's the initial code we went over yesterday, plus loading secrets from a file.  It said I didn't have write permissions yet, so I created a fork.  I'll continue working on the items we talked about.

### CODE REVIEW

**Comment by eneemann** [comment]
@steveoh I repointed this to the dev branch

File: cool.py [hunk 1]
```diff
@@ -97,18 +97,12 @@ def download_tiles(col, row):
     #: dowload bottom-right tile
     img_4 = requests.get(f"{BASE_URL}/{col_num + 1}/{row_num + 1}").content
 
-    tile_images = [img_1, img_2, img_3, img_4]
+    tile_list = [img_1, img_2, img_3, img_4]
```
Comment (steveoh) [inline cool.py:100]:
```suggestion
    return [img_1, img_2, img_3, img_4]
```

File: cool.py [hunk 2]
```diff
@@ -97,18 +97,12 @@ def download_tiles(col, row):
...
 
-    cv2_images = []
+    return tile_list
```
Comment (steveoh) [inline cool.py:102]:
```suggestion
```

**Review by steveoh — COMMENTED** [review]
**suggestion**: remove unnecessary variable

File: cool.py [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+
+if "PY_ENV" in environ and environ["PY_ENV"] == "production":
+    LOGGING_CLIENT = google.cloud.logging.Client()
+    STORAGE_CLIENT = google.cloud.storage.Client()
+
+    LOGGING_CLIENT.setup_logging()
+
+QUAD_WORD = None
+MODEL = None
+secrets = None
```
Comment (steveoh) [inline cool.py:32]:
```suggestion
SECRETS = None
```

File: cool.py [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+        col (str): the column of the WMTS index for the tile of interest (top-left tile)
+        row (str): the row of the WMTS index for the tile of interest (top-left tile)
+    Returns:
+        cv2_images (list): list of cv2 images
+    """
+    global secrets
+
+    if secrets is None:
+        logging.info('loading secrets')
+        secrets = SimpleNamespace(**_get_secrets())
```
Comment (steveoh) [inline cool.py:85]: **praise**: great use of simple namespace!

File: cool.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+    Args:
+        None
+
+    Returns:
+        pytorch model: model ready for scanning
+    """
+    model_weight_path = Path(__file__).parent / 'tower_scout' / 'xl_250_best.pt'
+    yolov5_path = Path(__file__).parent / 'yolov5'
+    model = torch.hub.load(str(yolov5_path), 'custom', path=str(model_weight_path), source='local')
+    #: if statements (check that .pt file exists, folders, etc.)
```
Comment (steveoh) [inline cool.py:178]: **suggestion**: yup. get defensive here in case people didn't run the prerequisites.

File: cool.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+    #     logging.info('working with a numpy array')
+    #     pil_image = Image.fromarray(image)
+
+    logging.info('working with %s', type(image))
+
+    scan_start = perf_counter()
+    results = towerscout_model(image)
+    logging.info('time to scan image: %s',  format_time(perf_counter() - scan_start))
+
+    results.save()
```
Comment (steveoh) [inline cool.py:224]: **issue**: this should be removed or put behind a cli flag somehow

File: cool.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+    #     pil_image = Image.open(image)
+    # elif isinstance(image, np.ndarray):
+    #     logging.info('working with a numpy array')
+    #     pil_image = Image.fromarray(image)
+
+    logging.info('working with %s', type(image))
+
+    scan_start = perf_counter()
+    results = towerscout_model(image)
+    logging.info('time to scan image: %s',  format_time(perf_counter() - scan_start))
```
Comment (steveoh) [inline cool.py:222]:
```suggestion
    logging.info('time to scan image: %s', format_time(perf_counter() - scan_start))
```

File: cool.py [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+    #: check for empty dataframe
+    if results_df.empty:
+        logging.info('empty dataframe, no cooling towers detected')
+
+        return results_df
+    
+    zoom_level = 20
+
+    #: get upper-left coordinates of the primary tile (upper-left tile)
+    tile = mercantile.ul(int(col), int(row), zoom_level)
```
Comment (steveoh) [inline cool.py:251]: **question**: what happens if col or row are not integers?

File: cool.py [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+
+    transformer = pyproj.Transformer.from_crs(pyproj.CRS.from_epsg(wgs84), pyproj.CRS.from_epsg(web_mercator), always_xy=True)
+    x, y = transformer.transform(tile.lng, tile.lat)
+
+    #: calculate centroid x/y coords in web mercator
+    meters_per_pixel = 0.1492910708688
+    results_df['x_centroid_3857'] = x + results_df['x_centroid_px'] * meters_per_pixel
+    results_df['y_centroid_3857'] = y - results_df['y_centroid_px'] * meters_per_pixel
+
+    print(results_df.head().to_string())
```
Comment (steveoh) [inline cool.py:269]: **chore**: remove debugging

**Review by steveoh — COMMENTED** [review]
please add the following items from the agrc/python repo

- `.github` folder
- `.vscode` folder
- `.coveragerc` file
- `.editorconfig` file
- `LICENSE` file
- `pylintrc` file
- `setup.cfg` file

File: cool.py (hunk 8 above)
Comment (eneemann) [reply cool.py:251]: it throws an error:
`TypeError: '<=' not supported between instances of 'int' and 'str'`

File: cool.py (hunk 1 above)
Comment (eneemann) [reply cool.py]: done

File: cool.py (hunk 2 above)
Comment (eneemann) [reply cool.py]: refactored this a bit

File: cool.py (hunk 3 above)
Comment (eneemann) [reply cool.py]: done

File: cool.py (hunk 6 above)
Comment (eneemann) [reply cool.py]: moved to CLI

File: cool.py (hunk 7 above)
Comment (eneemann) [reply cool.py]: done

File: cool.py (hunk 9 above)
Comment (eneemann) [reply cool.py]: done

File: cool_cli.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+Options:
+    --from=location                 The bucket or directory to operate on
+    --task-index=index              The index of the task running
+    --instances=size                The number of containers running the job [default: 10]
+    --save-to=location              The location to output the stuff
+Examples:
+    python cool_cli.py download tiles 198259 394029 --save-to=./data --mosaic
+    python cool_cli.py download tiles 198402 394067 --mosaic --detect
+    python cool_cli.py download tiles 198402 394067 --mosaic --detect --parse-results
```
[the remark this answers is not in the document]
Comment (eneemann) [reply cool_cli.py]: I'll let you review the latest changes

**Comment by eneemann** [comment]
Okay @steveoh, tear it apart. I'm kinda stuck on how to organize the CLI.  It doesn't make sense to me to run the mosaic command alone, because we'd have to build a list of images and read them in or refactor the function, so I packaged that as an option on download. I changed parse-results to locate-results because that function is creating the spatial info (but we need to specify the col/row to enable a mosaic image to be located).  The process-tile command runs through the whole chain to this point.

I made some of the changes you requested so far. It took a while, but I think I've refactored the tile downloads into something more robust with timeouts and retries, borrowing a bunch from Jake's Palletjack code.

For the BGR/RGB issue, it was only affecting in-memory np.ndarrays, so I created a function to reorder the color channels when a np.ndarray is detected. I think the cv2 function knows it has to write out as RGB when saving JPGs, so pushing a Path object for the JPG through the PyTorch model wasn't impacted by the reversed colors.

Still more to do (including tests), but making a little progress.

File: cool.py [hunk 11]
```diff
@@ -65,41 +64,165 @@ def convert_to_cv2_image(image):
...
+    #: the tile request is successful
+    if response.status_code == 200:
+        return response
+
+    #: the tile request fails
+    if response.status_code == 404:
+        return None
+
+    #: if we haven't returned, raise an error to trigger _retry
+    raise RuntimeError(f"Did not receive a valid tile download response; status code: {response.status_code}")
```
Comment (steveoh) [inline cool.py:160]: https://requests.readthedocs.io/en/latest/api/#requests.Response.raise_for_status

Comment (steveoh) [reply cool.py:160]: I prefer scott's masquerade code for [retries](https://github.com/agrc/masquerade/blob/main/src/masquerade/providers/web_api.py#L67) as it uses built it library code. His `make_request` method could use some exception handling e.g. 

```py
import requests

try:
    response = requests.get(url, timeout=5)
    response.raise_for_status() # Raise an exception if the status code is not 200 OK
except requests.exceptions.Timeout:
    # Timeout occurred
    print("Timeout error occurred")
except requests.exceptions.RequestException as e:
    # Other error occurred
    print("Error occurred: {}".format(e))
```

But if you were to combine the two I think you'd have a good reference implementation that masquerade and palletjack could benefit from.

File: cool_cli.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (steveoh) [reply cool_cli.py]: Yeah I like what you did.

File: cool.py [hunk 12]
```diff
@@ -141,23 +131,31 @@ def get_tile(url):
         dict: The 'results' dictionary of the response json (location, score, and matchAddress)
     """
 
-    response = requests.get(url, timeout=5)
+    session = _get_retry_session()
+
+    try:
+        response = session.get(url, timeout=5)
+        response.raise_for_status()  #: raise an exception if the status code is not 200 OK
```
Comment (steveoh) [inline cool.py:138]: **thought**: you might need to be careful here as a 404 will throw but we probably just want to skip it from the mosaic unless all of the tiles 404 and you just have a white image.

File: cool.py [hunk 13]
```diff
@@ -141,23 +131,31 @@ def get_tile(url):
...
+    except requests.exceptions.Timeout:
+        #: timeout occurred
+        logging.debug("Timeout error occurred")
```
Comment (steveoh) [inline cool.py:141]: **chore**: elevate this to an error or at least a warning.

File: cool.py [hunk 14]
```diff
@@ -141,23 +131,31 @@ def get_tile(url):
...
+
+        return None
+
+    except requests.exceptions.RequestException as e:
+        #: other error occurred
+        logging.debug("Error occurred: %s", e)
```
Comment (steveoh) [inline cool.py:147]: **chore**: elevate this to a warning or higher

File: cool.py (hunk 8 above)
Comment (eneemann) [reply cool.py:375]: For this one, we're already casting the col/row values as ints in the function call.  Only strings of integers will be read into the row/col variables from our BQ table.  So what else should be done to be more defensive?

Comment (steveoh) [reply cool.py:375]: Generally speaking it's unsafe to assume that a database or user input will always be what you expect. This is the first time in this function the variables are cast to int. If it's not the the first time in the flow then the cast is unnecessary as you can ensure they are ints already. It's trivial to have a type check or a try except and exit early so you don't bomb out with an unhandled ValueError.

**Pull request merged**